src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/ExceptionRegionEncoder.cs

```diff
@@ -2,120 +2,145 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+
 namespace System.Reflection.Metadata.Ecma335
 {
     public struct ExceptionRegionEncoder
     {
-        private readonly int _exceptionRegionCount;
+        private const int TableHeaderSize = 4;
+
+        private const int SmallRegionSize =
+            sizeof(short) +  // Flags
+            sizeof(short) +  // TryOffset
+            sizeof(byte) +   // TryLength
+            sizeof(short) +  // HandlerOffset
+            sizeof(byte) +   // HandleLength
+            sizeof(int);     // ClassToken | FilterOffset
+
+        private const int FatRegionSize =
+            sizeof(int) +    // Flags
+            sizeof(int) +    // TryOffset
+            sizeof(int) +    // TryLength
+            sizeof(int) +    // HandlerOffset
+            sizeof(int) +    // HandleLength
+            sizeof(int);     // ClassToken | FilterOffset
+
+        public BlobBuilder Builder { get;}
         private readonly bool _isSmallFormat;
-        public BlobBuilder Builder { get; }
 
-        internal ExceptionRegionEncoder(BlobBuilder builder, int exceptionRegionCount, bool hasLargeRegions)
+        internal ExceptionRegionEncoder(BlobBuilder builder, bool isSmallFormat)
         {
             Builder = builder;
-            _exceptionRegionCount = exceptionRegionCount;
-            _isSmallFormat = !hasLargeRegions && IsSmallRegionCount(exceptionRegionCount);
+            _isSmallFormat = isSmallFormat;
         }
 
-        public void StartRegions()
+        public static bool IsSmallRegionCount(int exceptionRegionCount) => 
+            GetExceptionTableSize(exceptionRegionCount, isSmallFormat: true) <= 0xff;
+
+        public static bool IsSmallExceptionRegion(int startOffset, int length) => 
+            startOffset <= 0xffff && length <= 0xff;
+
+        internal static bool IsSmallExceptionRegionFromBounds(int startOffset, int endOffset) => 
+            IsSmallExceptionRegion(startOffset, endOffset - startOffset);
+
+        internal static int GetExceptionTableSize(int exceptionRegionCount, bool isSmallFormat) => 
+            TableHeaderSize + exceptionRegionCount * (isSmallFormat ? SmallRegionSize : FatRegionSize);
+
+        internal static bool IsExceptionRegionCountInBounds(int exceptionRegionCount) => 
+            TableHeaderSize + (long)unchecked((uint)exceptionRegionCount) * FatRegionSize <= 0xffffff; 
+        
```

Comment:
Ah, of course of them (0xffffff) is none of the above. Separate constant for that would be nice.